hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java

```diff
@@ -849,6 +856,25 @@ private void validateSnapshotsExistAndActive(final String volumeName,
     checkSnapshotActive(toSnapInfo, false);
   }
 
+  public void validateSnapshotLimit() throws IOException {
```

Comment:
Make this method synchronized and have a parameter in memory which keeps the number of snapshots creation in flight in memory. In hindsight we should not put this check inside validateAndUpdateCache since this can impact ratis replays. @hemantk-12 
The in memory count should be increased once the pre execute passes inside a synchronized block and should be decremented once the snapshot gets added on to the snapshot chain in validateAndUpdate cache. The total number of snapshots at any time would be:
totalSnapshots = snapshotChainManager.size() + inmemoryCounter

Reply:
To add more details:
Ratis replay would not be deterministic across various om nodes if the config is not matching across all nodes. We should only look at leader. Since validateAndUpdateCache runs on all nodes, this could lead to an inconsistent state.

Reply:
Rename this function addSnapshotLimitCheck() or something

Reply:
This approach has an issue. `preExecute` runs only on a single node (the leader node, if I recall correctly), whereas `validateAndUpdateCache` runs on all nodes. If we decrement in `validateAndUpdateCache`, other nodes will keep decrementing without a corresponding increment, causing the value to go negative. Additionally, if leadership changes, it could result in exceeding the allowed limit.

IMO, we can simplify this by relying on the `SnapshotChain` count, which will act as a soft limit initially and eventually become a hard limit. I doubt that `applyingTransaction` is slower than the write request rate to the extent that it would create a backlog of snapshot creation requests.

Reply:
Yes, we can ignore the inflight counter on the followers and start from `0`, but if there is a leadership change while `preExecute` has completed and `validateAndUpdateCache` has not. When a new leader (former follower) tries to apply the transactions (basically executing `validateAndUpdateCache`), `inflight` counter will go negative.

Reply:
@swamirishi and I were discussing this, and one way to fix it is to implement `notifyLeaderReady()` and `notifyNotLeader()` similar to [SCMStateMachine](https://github.com/apache/ozone/blob/7dfd8c1a7889b5947c87f81d1c4d24a30bf629f6/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java#L372) for OzoneManager and reset the `inflight` counter on leadership change.

cc: @swamirishi @SaketaChalamchala